# Overview of analysis steps used to generate results used for manuscript

## Temp section
* Important files?
* Alignment using ngmlr
  * done
* generating tdf file
* pbsv discover
* pbsv call
* Filtering SVs
* Finding somatic mutations

## Overview
* The goal of this file is to provide all the analysis steps used to generate \
the results in the updated Poplar genome paper in the works
* Most/all of the steps are detailed in other .md files; I've done my best \
to transfer all the necessary and important information here, but if there \
are questions, it will probably help to do some sleuthing in those other files.
* There were many attempts and directions that I started but did not use in \
for the final analysis - details can be found in the other .md files

## List of important files
* Metadata

## Alignment of PacBio reads using ngmlr
### Overview
* Merged PacBio fasta.gz files from different cells to make single fasta.gz \
file for each branch
* Ran ngmlr for each branch
* Sorted .bam files generated by ngmlr
### Gather and concatenate fasta files
#### Test approach using one library/sample: PBAU
##### Location of .fasta files for PBAU
```
/home/raid2/PACBIOPROC/Sequel/PBmixSequel392/4_F01
/home/raid2/PACBIOPROC/Sequel/PBmixSequel392/5_G01
/home/raid2/PACBIOPROC/Sequel/PBmixSequel392/6_H01
/home/raid2/PACBIOPROC/Sequel/PBmixSequel389/2_B01

```
##### Command
```
cat /home/raid2/PACBIOPROC/Sequel/PBmixSequel392/4_F01/m54037_180609_042425.subreads.fasta.gz /home/raid2/PACBIOPROC/Sequel/PBmixSequel392/5_G01/m54037_180609_143801.subreads.fasta.gz /home/raid2/PACBIOPROC/Sequel/PBmixSequel392/6_H01/m54037_180610_005237.subreads.fasta.gz /home/raid2/PACBIOPROC/Sequel/PBmixSequel389/2_B01/m54037_180607_065358.subreads.fasta.gz > PBAU_merged_fasta.gz

```
#### Use metadata to batch-run concatenating fasta files for the libraries 
##### Metadata files
* Sample metadata
  * `/home/t4c1/WORK/grabowsk/data/poplar_branches/meta/poplar_branch_meta_v3.0.txt`
* Sequencer metadata including directory of .fasta files
  * `/home/t4c1/WORK/grabowsk/data/poplar_branches/meta/pop_branch_sequencer_info.tsv`
##### R script
R script to use sequencing and sample metadata to creat `cat` commands to
merge/concatenate the fasta files from each sample/library
```
./r_scripts/lib_fasta_cat_commands.r
```
##### shell script
Generated by the R script
```
/home/f1p1/tmp/poplar_branches/lib_mapping/cat_popbranch_subread_fastas.sh
```
##### merged fasta files
* found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* named `LIBNAME_merged_fasta.gz`, with LIBNAME replaced with name of the library
  * ex: `PAXL_merged_fasta.gz`
### Run ngmlr on merged fasta files
#### Overview
* run ngmlr and pipe into samtools to generate .bam file
* need to set appropriate flag in ngmlr, `--bam-fix`, so that output can be\
processed by samtools
* I chose to multithread using 32 threads - this takes up one full 32-thread\
machine on `all.q`
#### Using conda environment
* submission script includes setting conda environment so has ngmlr, samtools,\
etc. loaded
  * `source /home/raid2/LINUXOPT/miniconda2/bin/activate /home/grabowsky/.conda/envs/NGS_analysis`
  * This environment was cloned from: `/home/raid2/LINUXOPT/AnacondaEnv/Human_Analysis/`
#### Testing with one library: PBAU
##### Shell script
```
/home/f1p1/tmp/poplar_branches/lib_mapping/PBAU.14.5v1.0Ref.ngmlr.sh
```
##### Submit script
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
qsub PBAU.14.5v1.0Ref.ngmlr.sh
```
Note: Worked! Took 9 hrs 18 min
#### Run on remaining libraries
##### Generate shell scripts
Generate submission scripts for other 9 libraries
```
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do sed 's/PBAU/'"$i"'/g' PBAU.14.5v1.0Ref.ngmlr.sh > $i.14.5v1.0Ref.ngmlr.sh ; done
```
##### Submit scripts
```
bash
cd /home/f1p1/tmp/poplar_branches/lib_mapping
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do qsub $i.14.5v1.0Ref.ngmlr.sh ; done
```
#### Output
* Produced .bam file for each sample
* Found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* Named `LIBNAME.14.5v1.0Ref.ngmlr.bam`
  * ex: `PAXL.14.5v1.0Ref.ngmlr.bam`
### Sort .bam Files
#### Overview
* Need to sort the .bam files before running SNIFFLES
#### Testing with one library: PBAU
##### Shell script
```
/home/f1p1/tmp/poplar_branches/lib_mapping/PBAU.14.5v1.0Ref.sortbam.sh
```
##### Submit job
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
qsub PBAU.14.5v1.0Ref.sortbam.sh
```
#### Run on Remaining Libraries
##### Generate shell scripts
```
bash
cd /home/f1p1/tmp/poplar_branches/lib_mapping
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do sed 's/PBAU/'"$i"'/g' PBAU.14.5v1.0Ref.sortbam.sh > $i.14.5v1.0Ref.sortbam.sh ; done
```
##### Submit jobs
```
bash
cd /home/f1p1/tmp/poplar_branches/lib_mapping
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do qsub $i.14.5v1.0Ref.sortbam.sh ; done
```
#### Output
* Produced sorted .bam file for each sample
* Found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* Named `LIBNAME.14.5v1.0Ref.ngmlr.sorted.bam`
  * ex: `PAXL.14.5v1.0Ref.ngmlr.sorted.bam`





