# Running ngmlr for mapping poplar branch PacBio data

## Goals
* Map poplar branch PacBio libraries to 14.5 v1.0 (based on Pop Ref 4.0)
* Detect SV in each library
* Genotype libraries at all SVs found in any library

## Steps
* Gather and concatenate fasta files
* ngmlr to map sample fasta files to Poplar 14.5 v1.0 reference to create .bam\
files
* sort .bam files
* SNIFFLES on each .bam individually to create vcf files
* sort vcf files
* SURVIVOR to find all SVs in all branches
* SNIFFLES a second time using SURVIVOR output as "map"
* sort vcf files
* SURVIVOR a second time to get combined VCF with all SVs genotyped in all\
samples

## Gather and concatenate fasta files
### Test approach using one library/sample: PBAU
#### Location of .fasta files for PBAU
```
/home/raid2/PACBIOPROC/Sequel/PBmixSequel392/4_F01
/home/raid2/PACBIOPROC/Sequel/PBmixSequel392/5_G01
/home/raid2/PACBIOPROC/Sequel/PBmixSequel392/6_H01
/home/raid2/PACBIOPROC/Sequel/PBmixSequel389/2_B01

```
#### Command
```
cat /home/raid2/PACBIOPROC/Sequel/PBmixSequel392/4_F01/m54037_180609_042425.subreads.fasta.gz /home/raid2/PACBIOPROC/Sequel/PBmixSequel392/5_G01/m54037_180609_143801.subreads.fasta.gz /home/raid2/PACBIOPROC/Sequel/PBmixSequel392/6_H01/m54037_180610_005237.subreads.fasta.gz /home/raid2/PACBIOPROC/Sequel/PBmixSequel389/2_B01/m54037_180607_065358.subreads.fasta.gz > PBAU_merged_fasta.gz

```
Note: Works!
### Use metadata to batch-run concatenating fasta files for the libraries 
#### Metadata files
* Sample metadata
  * `/home/t4c1/WORK/grabowsk/data/poplar_branches/meta/poplar_branch_meta_v3.0.txt`
* Sequencer metadata including directory of .fasta files
  * `/home/t4c1/WORK/grabowsk/data/poplar_branches/meta/pop_branch_sequencer_info.tsv`
#### R script
R script to use sequencing and sample metadata to creat `cat` commands to
merge/concatenate the fasta files from each sample/library
```
./r_scripts/lib_fasta_cat_commands.r
```
#### shell script
Generated by the R script
```
/home/f1p1/tmp/poplar_branches/lib_mapping/cat_popbranch_subread_fastas.sh
```
#### merged fasta files
* found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* named `LIBNAME_merged_fasta.gz`, with LIBNAME replaced with name of the library
  * ex: `PAXL_merged_fasta.gz`

## Run ngmlr on merged fasta files
### Overview
* run ngmlr and pipe into samtools to generate .bam file
* need to set appropriate flag in ngmlr, `--bam-fix`, so that output can be\
processed by samtools
* I chose to multithread using 32 threads - this takes up one full 32-thread\
machine on `all.q` 
### Using conda environment
* submission script includes setting conda environment so has ngmlr, samtools,\
etc. loaded
  * `source /home/raid2/LINUXOPT/miniconda2/bin/activate /home/grabowsky/.conda/envs/NGS_analysis`
  * This environment was cloned from: `/home/raid2/LINUXOPT/AnacondaEnv/Human_Analysis/`
### Testing with one library: PBAU
#### Shell script
```
/home/f1p1/tmp/poplar_branches/lib_mapping/PBAU.14.5v1.0Ref.ngmlr.sh
```
#### Submit script
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
qsub PBAU.14.5v1.0Ref.ngmlr.sh
```
Note: Worked! Took 9 hrs 18 min
### Run on remaining libraries
#### Generate shell scripts
Generate submission scripts for other 9 libraries
```
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do sed 's/PBAU/'"$i"'/g' PBAU.14.5v1.0Ref.ngmlr.sh > $i.14.5v1.0Ref.ngmlr.sh ; done
```
#### Submit scripts
```
bash
cd /home/f1p1/tmp/poplar_branches/lib_mapping
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do qsub $i.14.5v1.0Ref.ngmlr.sh ; done
```
### Output
* Produced .bam file for each sample
* Found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* Named `LIBNAME.14.5v1.0Ref.ngmlr.bam`
  * ex: `PAXL.14.5v1.0Ref.ngmlr.bam`

## Sort .bam Files
Need to sort the .bam files before running SNIFFLES

