# Running ngmlr for mapping poplar branch PacBio data

## Goals
* Map poplar branch PacBio libraries to 14.5 v1.0 (based on Pop Ref 4.0)
* Detect SV in each library
* Genotype libraries at all SVs found in any library

## Steps
* Gather and concatenate fasta files
* ngmlr to map sample fasta files to Poplar 14.5 v1.0 reference to create .bam\
files
* sort .bam files
* SNIFFLES on each .bam individually to create vcf files
* sort vcf files
* SURVIVOR to find all SVs in all branches
* SNIFFLES a second time using SURVIVOR output for supervised mapping
* sort vcf files
* SURVIVOR a second time to get combined VCF with all SVs genotyped in all\
samples

## Gather and concatenate fasta files
### Test approach using one library/sample: PBAU
#### Location of .fasta files for PBAU
```
/home/raid2/PACBIOPROC/Sequel/PBmixSequel392/4_F01
/home/raid2/PACBIOPROC/Sequel/PBmixSequel392/5_G01
/home/raid2/PACBIOPROC/Sequel/PBmixSequel392/6_H01
/home/raid2/PACBIOPROC/Sequel/PBmixSequel389/2_B01

```
#### Command
```
cat /home/raid2/PACBIOPROC/Sequel/PBmixSequel392/4_F01/m54037_180609_042425.subreads.fasta.gz /home/raid2/PACBIOPROC/Sequel/PBmixSequel392/5_G01/m54037_180609_143801.subreads.fasta.gz /home/raid2/PACBIOPROC/Sequel/PBmixSequel392/6_H01/m54037_180610_005237.subreads.fasta.gz /home/raid2/PACBIOPROC/Sequel/PBmixSequel389/2_B01/m54037_180607_065358.subreads.fasta.gz > PBAU_merged_fasta.gz

```
Note: Works!
### Use metadata to batch-run concatenating fasta files for the libraries 
#### Metadata files
* Sample metadata
  * `/home/t4c1/WORK/grabowsk/data/poplar_branches/meta/poplar_branch_meta_v3.0.txt`
* Sequencer metadata including directory of .fasta files
  * `/home/t4c1/WORK/grabowsk/data/poplar_branches/meta/pop_branch_sequencer_info.tsv`
#### R script
R script to use sequencing and sample metadata to creat `cat` commands to
merge/concatenate the fasta files from each sample/library
```
./r_scripts/lib_fasta_cat_commands.r
```
#### shell script
Generated by the R script
```
/home/f1p1/tmp/poplar_branches/lib_mapping/cat_popbranch_subread_fastas.sh
```
#### merged fasta files
* found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* named `LIBNAME_merged_fasta.gz`, with LIBNAME replaced with name of the library
  * ex: `PAXL_merged_fasta.gz`

## ngmlr Steps
### Run ngmlr on merged fasta files
#### Overview
* run ngmlr and pipe into samtools to generate .bam file
* need to set appropriate flag in ngmlr, `--bam-fix`, so that output can be\
processed by samtools
* I chose to multithread using 32 threads - this takes up one full 32-thread\
machine on `all.q` 
#### Using conda environment
* submission script includes setting conda environment so has ngmlr, samtools,\
etc. loaded
  * `source /home/raid2/LINUXOPT/miniconda2/bin/activate /home/grabowsky/.conda/envs/NGS_analysis`
  * This environment was cloned from: `/home/raid2/LINUXOPT/AnacondaEnv/Human_Analysis/`
#### Testing with one library: PBAU
##### Shell script
```
/home/f1p1/tmp/poplar_branches/lib_mapping/PBAU.14.5v1.0Ref.ngmlr.sh
```
##### Submit script
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
qsub PBAU.14.5v1.0Ref.ngmlr.sh
```
Note: Worked! Took 9 hrs 18 min
#### Run on remaining libraries
##### Generate shell scripts
Generate submission scripts for other 9 libraries
```
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do sed 's/PBAU/'"$i"'/g' PBAU.14.5v1.0Ref.ngmlr.sh > $i.14.5v1.0Ref.ngmlr.sh ; done
```
##### Submit scripts
```
bash
cd /home/f1p1/tmp/poplar_branches/lib_mapping
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do qsub $i.14.5v1.0Ref.ngmlr.sh ; done
```
#### Output
* Produced .bam file for each sample
* Found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* Named `LIBNAME.14.5v1.0Ref.ngmlr.bam`
  * ex: `PAXL.14.5v1.0Ref.ngmlr.bam`

### Sort .bam Files
#### Overview
* Need to sort the .bam files before running SNIFFLES
#### Testing with one library: PBAU
##### Shell script
```
/home/f1p1/tmp/poplar_branches/lib_mapping/PBAU.14.5v1.0Ref.sortbam.sh
```
##### Submit job
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
qsub PBAU.14.5v1.0Ref.sortbam.sh
```
#### Run on Remaining Libraries
##### Generate shell scripts
```
bash
cd /home/f1p1/tmp/poplar_branches/lib_mapping
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do sed 's/PBAU/'"$i"'/g' PBAU.14.5v1.0Ref.sortbam.sh > $i.14.5v1.0Ref.sortbam.sh ; done
```
##### Submit jobs
```
bash
cd /home/f1p1/tmp/poplar_branches/lib_mapping
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do qsub $i.14.5v1.0Ref.sortbam.sh ; done
```
#### Output
* Produced sorted .bam file for each sample
* Found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* Named `LIBNAME.14.5v1.0Ref.ngmlr.sorted.bam`
  * ex: `PAXL.14.5v1.0Ref.ngmlr.sorted.bam`

## SNIFFLES Pipeline (first runthrough)
### SNIFFLES First Run
#### Overview
* Run SNIFFLES on sorted .bam files
* Generates .vcf file with SV genotypes. VCF includes info about the SVs in \
the INFO and GT fields
* Used following flags:
  * -s 5
  * -l 30
    * originally used 50 but decided to use 30 and then filter out those \
below 50bp
  * -n 5
#### Testing with one library: PBAU
##### Shell script
```
/home/f1p1/tmp/poplar_branches/lib_mapping/PBAU.14.5v1.0Ref.sorted.shiffles.sh
``` 
##### Submit job
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
qsub PBAU.14.5v1.0Ref.sorted.shiffles.sh
```
#### Run on Remaining Libraries
##### Generate shell scripts
```
bash
cd /home/f1p1/tmp/poplar_branches/lib_mapping
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do sed 's/PBAU/'"$i"'/g' PBAU.14.5v1.0Ref.sorted.shiffles.sh > $i.14.5v1.0Ref.sorted.shiffles.sh ; done
```
##### Submit jobs
```
bash
cd /home/f1p1/tmp/poplar_branches/lib_mapping
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do qsub $i.14.5v1.0Ref.sorted.shiffles.sh ; done
```
#### Output
* Produces .vcf with genotypes for SVs
* Found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* Named `LIBNAME.14.5v1.0Ref.ngmlr.sorted.sniffles.vcf`
  * ex: `PAXL.14.5v1.0Ref.ngmlr.sorted.sniffles.vcf`

### Sort .vcf files
#### Overview
* Sort .vcf output using the `vcf-sort` Perl script from VCFtools
* SURVIVOR requires the .vcf files to be sorted before comparing files
#### Testing with one library: PBAU
```
cat PBAU.14.5v1.0Ref.ngmlr.sorted.sniffles.vcf | vcf-sort > PBAU.14.5v1.0Ref.ngmlr.sorted.sniffles.sorted.vcf
```
#### Run on Remaining Libraries
```
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do cat $i.14.5v1.0Ref.ngmlr.sorted.sniffles.vcf | vcf-sort > $i.14.5v1.0Ref.ngmlr.sorted.sniffles.sorted.vcf; done
```
#### Output
* Produces sorted .vcf
* Found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* Named `LIBNAME.14.5v1.0Ref.ngmlr.sorted.sniffles.sorted.vcf`
  * ex: `PAXL.14.5v1.0Ref.ngmlr.sorted.sniffles.sorted.vcf`

### SURVIVOR with Raw Genotypes
#### Overview
* First generate file with names of single sample .vcf files to be included in \
merged .vcf
* SURVIVOR generates .vcf with genotypes for all the samples used in the \
analysis
  * Many genotypes will be incomplete because of low coverage or lack of \
variation in certain samples
* Used following values/flags: `1000 1 1 -1 -1 -1`
#### Generate File with Names of Single Sample VCFs
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
ls *sniffles.sorted.vcf > pop_sniffles_vcf_files_raw_call.txt
```
#### Run SURVIVOR
##### Shell Script
```
/home/f1p1/tmp/poplar_branches/lib_mapping/popbranches.survivor.merge.raw.vcfs.sh
```
##### Run job
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
qsub popbranches.survivor.merge.raw.vcfs.sh
```
#### Output
* Generates merged .vcf with genotypes for all the samples
* Found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* File name: `popbranch.ngmlr.sniffles.survivor.rawmerged1kbdist.vcf`

### Supervised SNIFFLES
#### Overview
* Use merged .vcf from SURVIVOR to supervise SV genotyping with SNIFFLES
#### Testing with one library: PBAU
##### Shell script
```
/home/f1p1/tmp/poplar_branches/lib_mapping/PBAU.14.5v1.0Ref.supervised.sniffles.sh
```
##### Submit job
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
qsub PBAU.14.5v1.0Ref.supervised.sniffles.sh
```
#### Run on Remaining Libraries
##### Generate shell scripts
```
bash
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do sed 's/PBAU/'"$i"'/g' PBAU.14.5v1.0Ref.supervised.sniffles.sh > $i.14.5v1.0Ref.supervised.sniffles.sh ; done
```
##### Run jobs
```
bash
cd /home/f1p1/tmp/poplar_branches/lib_mapping
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do qsub $i.14.5v1.0Ref.supervised.sniffles.sh ; done
```
#### Output
* Generates .vcfs
* Found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* Named `LIBNAME.14.5v1.0Ref.ngmlr.sorted.sniffles_supervised.vcf`
  * ex: `PAXL.14.5v1.0Ref.ngmlr.sorted.sniffles_supervised.vcf`

### Sort Supervised VCFs
#### Overview
* Sort the supervised .vcfs using the `vcf-sort` Perl script from VCFtools
* Need to sort the .vcfs for the next step with SURVIVOR
#### Testing with one library: PBAU
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
cat PBAU.14.5v1.0Ref.ngmlr.sorted.sniffles_supervised.vcf | vcf-sort > PBAU.14.5v1.0Ref.ngmlr.sorted.sniffles_sup.sorted.vcf
```
#### Run on remaining libraries
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do cat $i.14.5v1.0Ref.ngmlr.sorted.sniffles_supervised.vcf | vcf-sort > $i.14.5v1.0Ref.ngmlr.sorted.sniffles_sup.sorted.vcf; done
```
#### Output
* Produces sorted .vcfs
* Found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* Named `LIBNAME.14.5v1.0Ref.ngmlr.sorted.sniffles_sup.sorted.vcf`
  * ex: `PAXL.14.5v1.0Ref.ngmlr.sorted.sniffles_sup.sorted.vcf`

### SURVIVOR to Generate Final Merged VCF
#### Overview
* Need to generate file with list of supervised, sorted .vcfs to be merged
* SURVIVOR generates .vcf with geneotypes for all sampels at each SV
* Used the following flags/values: `1000 0 1 -1 -1 -1`
#### Generate File with List of supervised, sorted .vcfs
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
ls *sniffles_sup.sorted.vcf > popbranch_sniffles_vcf_files_supervised.txt
```
#### Run job
##### Shell script
```
/home/f1p1/tmp/poplar_branches/lib_mapping/popbranches.survivor.merge.supervised.vcfs.sh
```
##### Submit job
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
qsub popbranches.survivor.merge.supervised.vcfs.sh

```
##### Output
* Generates merged .vcf with genotypes at all SVs for all samples
* Found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* File name: `popbranch.ngmlr.sniffles.survivor.supervisedmerged1kbdist.vcf`

### Transfer VCFs to $WORK directory
#### Overview
* Transfer final merged VCF from SURVIVOR and raw individual VCFs to work \
directory for analysis
#### Commands
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping

for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAU PBAW; do cp $i'.14.5v1.0Ref.ngmlr.sorted.sniffles.vcf' /home/t4c1/WORK/grabowsk/data/poplar_branches/SV_calling_analysis/sniffles; done

cp popbranch.ngmlr.sniffles.survivor.supervisedmerged1kbdist.vcf /home/t4c1/WORK/grabowsk/data/poplar_branches/SV_calling_analysis/sniffles

```
#### Output
* Files found in following $WORK subdirectory: `/home/t4c1/WORK/grabowsk/data/poplar_branches/SV_calling_analysis/sniffles/`

## Generate Files for Fritz (developer) to troubleshoot
* Top 250 lines
* Files:
  * Initial VCF
  * Initial merged VCF
  * Supervised VCF
  * Final merged VCF

### Commands
#### Initial VCFs
* `PBAW.14.5v1.0Ref.ngmlr.sorted.sniffles.sorted.vcf`
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping

for LIBNAME in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAU PBAW; \
do head -250 $LIBNAME'.14.5v1.0Ref.ngmlr.sorted.sniffles.sorted.vcf' > \
$LIBNAME'.initialVCF_top250.vcf'; done

for LIBNAME in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAU PBAW; \
do head -250 $LIBNAME'.14.5v1.0Ref.ngmlr.sorted.sniffles_sup.sorted.vcf'> \
$LIBNAME'.supervisedVCF_top250.vcf'; done

head -250 popbranch.ngmlr.sniffles.survivor.rawmerged1kbdist.vcf > \
initial_merged_VCF_top250.vcf

head -250 popbranch.ngmlr.sniffles.survivor.supervisedmerged1kbdist.vcf > \
supervised_merged_VCF_top250.vcf

tar -cvzf poplar_files_top250lines.tar.gz *top250.vcf

```

## Re-run SURVIVOR on Raw Genotype VCFs
### Overview
* First generate file with names of single sample .vcf files to be included in \
merged .vcf
* SURVIVOR generates .vcf with genotypes for all the samples used in the \
analysis
  * Many genotypes will be incomplete because of low coverage or lack of \
variation in certain samples
* Used following values/flags: `1000 1 1 1 0 10`
### Generate File with Names of Single Sample VCFs
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
ls *sniffles.sorted.vcf > pop_sniffles_vcf_files_raw_call.txt
```
### Run SURVIVOR
#### Shell Script
```
/home/f1p1/tmp/poplar_branches/lib_mapping/popbranches.survivor.merge.raw.vcfs_Try2.sh
```
#### Run job
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
qsub popbranches.survivor.merge.raw.vcfs_Try2.sh
```
### Output
* Generates merged .vcf with genotypes for all the samples
* Found in `/home/f1p1/tmp/poplar_branches/lib_mapping`
* File name: `popbranch.ngmlr.sniffles.survivor.rawmerged1kbdist_Try2.vcf`

### Re-run supervised SNIFFLES using new merged VCF
#### Overview
* Use merged .vcf from updated SURVIVOR to supervise SV genotyping with SNIFFLES
#### Testing with one library: PBAU
##### Shell script
```
/home/f1p1/tmp/poplar_branches/lib_mapping/PBAU.14.5v1.0Ref.supervised.sniffles_Try2.sh
```
##### Submit job
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
qsub PBAU.14.5v1.0Ref.supervised.sniffles_Try2.sh
```
#### Run on Remaining Libraries
##### Generate shell scripts
```
bash
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do sed 's/PBAU/'"$i"'/g' PBAU.14.5v1.0Ref.supervised.sniffles_Try2.sh > $i.14.5v1.0Ref.supervised.sniffles_Try2.sh ; done
```
### SIDE-WAYS
* Check if get different output from first round of sniffles using newer \
version of sniffles
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
qsub PBAU.14.5v1.0Ref.sorted.shiffles_Try2.sh

```
### Course correct
* Decided at this point (Jan 2019) to rerun the SNIFFLES/SURVIVOR pipeline \
with the newest versions of both programs to see how the work and generate \
the issues Lori and I are seeing based on the updated versions
* I suspect I'll see the same issue as before with read-support falling off \
drasitcally when running SNIFFLES with the merged .vcf as input

## Jan 2019 - Re-run sniffles pipeline (`rt_v2` = runthrough version 2)
### Run SNIFFLES on .bam files
#### Overview
* Use SNIFFLES installed on Jan. 24 2019
* Run SNIFFLES on sorted .bam files generated in Oct 2018
* Generates .vcf file with SV genotypes. VCF includes info about the SVs in \
the INFO and GT fields
* Used following flags:
  * -s 5
  * -l 30
  * -n 5
#### Adjust old .sh to use updated version of SNIFFLES and new directory
```
cd /home/f1p1/tmp/poplar_branches/lib_mapping
cp PBAU.14.5v1.0Ref.sorted.shiffles_Try2.sh ../sniffles_rt_v2/
cd /home/f1p1/tmp/poplar_branches/sniffles_rt_v2
mv PBAU.14.5v1.0Ref.sorted.shiffles_Try2.sh PBAU.14.5v1.0Ref.sorted.sniffles_rt_v2.sh
```
* Then adjust commands in the .sh file
##### Generate .sh files for all libraries
```
bash
cd /home/f1p1/tmp/poplar_branches/sniffles_rt_v2
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; do sed 's/PBAU/'"$i"'/g' PBAU.14.5v1.0Ref.sorted.sniffles_rt_v2.sh > $i.14.5v1.0Ref.sorted.sniffles_rt_v2.sh ; done
```
#### Submit jobs
```
bash
cd /home/f1p1/tmp/poplar_branches/sniffles_rt_v2
for i in `ls *sorted.sniffles_rt_v2.sh`; do qsub $i; done
```
#### Output
* Produces .vcf with genotypes for SVs
* Found in `/home/f1p1/tmp/poplar_branches/sniffles_rt_v2`
* Named `LIBNAME.14.5v1.0Ref.ngmlr.sorted.sniffles_rt_v2.vcf`
  * ex: `PAXL.14.5v1.0Ref.ngmlr.sorted.sniffles_rt_v2.vcf`
### Sort raw .vcf files
#### Overview
* Sort .vcf output using the `vcf-sort` Perl script from VCFtools
* SURVIVOR requires the .vcf files to be sorted before comparing files
* Use VCFtools in conda environment to make sure is most up to date
#### Run on all libraries Libraries
```
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAU PBAW; do cat $i.14.5v1.0Ref.ngmlr.sorted.sniffles_rt_v2.vcf | vcf-sort > $i.14.5v1.0Ref.ngmlr.sorted.sniffles.sorted_rt_v2.vcf; done
```
### Output
* Produces sorted .vcf
* Found in `/home/f1p1/tmp/poplar_branches/sniffles_rt_v2/`
* Named `LIBNAME.14.5v1.0Ref.ngmlr.sorted.sniffles.sorted_rt_v2.vcf`
  * ex: `PAXL.14.5v1.0Ref.ngmlr.sorted.sniffles.sorted_rt_v2.vcf`

### Merge sorted raw .vcf files with SURVIVOR
#### Overview
* First generate file with names of single sample .vcf files to be included \
in merged .vcf
* Used following flags: `1000 1 1 1 0 10`
#### Generate file with names of .vcfs
```
cd /home/f1p1/tmp/poplar_branches/sniffles_rt_v2
ls *sniffles.sorted_rt_v2.vcf > pop_sniffles_VCFfiles_raw_rt_v2.txt
```
#### Run SURVIVOR
##### Shell script
###### Copy and adjust previous .sh script
```
cd /home/f1p1/tmp/poplar_branches/sniffles_rt_v2
cp ../lib_mapping/popbranches.survivor.merge.raw.vcfs_Try2.sh \
./pop.survivor.merge.raw.vcfs_rt_v2.sh
```
* Adjust code
#### Run job
```
cd /home/f1p1/tmp/poplar_branches/sniffles_rt_v2
qsub pop.survivor.merge.raw.vcfs_rt_v2.sh
```
#### Output
* In `/home/f1p1/tmp/poplar_branches/sniffles_rt_v2/`
* `popbranch.ngmlr.sniffles.survivor.rawmerged1kbdist_rt_v2.vcf`
### Run SNIFFLES using merged .vcf as input
#### Overview
* use merged .vcf as input for supervised VCF calling with SNIFFLES
#### Generate shell scripts
##### Adjust raw .sh script
```
cd /home/f1p1/tmp/poplar_branches/sniffles_rt_v2
cp PBAU.14.5v1.0Ref.sorted.sniffles_rt_v2.sh \
PBAU.14.5v1.0Ref.sorted.sniffles.supervised_rt_v2.sh
```
* Adjust commands
##### Create remaining shell files
```
bash
for i in PAXL PAXN PAYK PAYZ PAZF PAZG PAZH PBAT PBAW; \
do sed 's/PBAU/'"$i"'/g' \
PBAU.14.5v1.0Ref.sorted.sniffles.supervised_rt_v2.sh > \
$i.14.5v1.0Ref.sorted.sniffles.supervised_rt_v2.sh ; done
```
#### Run jobs
```
bash
cd /home/f1p1/tmp/poplar_branches/sniffles_rt_v2
for i in `ls *sorted.sniffles.supervised_rt_v2.sh`; do qsub $i; done
```
### Sort supervised .vcf files
### merge sored supervised .bcf files with SURVIVOR




