# Notes about filtered SVs from PacBio Joint SV Calling

## Overview
### General Info
* SVs generated by the PacBio pipeline are sort of wonky and need filtering
* Filter the SVs using a variety of criteria:
  * Type of SV
  * SV sequence content (repeats)
  * Proximity to other SVs
  * Quality of genotype scores
* Generate genotypes using custom script to generate high-quality genotypes
* Multi-step process to generate final genotype set:
  * Filtering SVs before calling genotypes (Pre-calling filtering)
  * Generating genotypes
  * Filtering SVs based on genotypes (Post-calling filtering)
### Files
* File with custom R scripts
  * `~/r_scripts/pb_SV_analysis_functions.r`
* PowerPoint with explanation of analysis and results
  * `~/ppt_files/SV_filtering_steps.pptx`

## Pre-Calling Filtering Steps
1. Select only insertions or deletions
  * Part of `make_combo_indel_df` function
  * Motivation is that INV and BND SVs don't work with downstream functions
2. Remove SVs with high mono- or binucleotide repeat content
  * `remove_8mer_SV_inds` function, though works with any k-mer length
3. Remove duplicate postions
  * `dup_inds_to_remove` function
  * retains the SV with the longest sequence at the postion
  * Motivation is that most are because different versions of the same SV are \
ID's in different samples
4. Remove nearby SVs
  * `overlap_inds_to_remove`
  * retains the SV with longest sequence
  * Motivation is that some overlapping SVs are the same SV called at \
slightly different positions in different samples so are not independent
    * Also, SVs should ge distributed somewhat equally across the genome so \
don't want small regions of the genome to bias the genome-wide analysis
5. Remove SVs with missing default genotypes
  * `remove_missing_combo_SVs` function
  * Specific to this analysis since all SV positions should be prsent in all \
clonal samples
6. Remove SVs with excessive coverage
  * `id_excess_coverage` function
  * uses SD-based cutoff as default, but can also use numerical coverage cutoff

## Genotype Calling
### Overview
* `call_allsamp_genotype` function
  * requires running `make_allsamp_geno_info_list` function ahead of time
  * is good to run the pre-calling filtering steps ahead of time
* Any genotypes that don't meet a requirement are assigned an NA genotype
### Genotype Requirements
* Minimum coverage to call genotype in an  individual
  * ex: 10 reads
* Minimum penetrance/read ratio to call heterozygous genotype
  * ex: 0.15
* Maximum penetrance/read ratio to call homozygous genotype
  * ex: 0.05
* Minimum number of minor allele reads in sample to call heterozygous genotype
  * ex: 2 reads
  * Necessary because depending on the coverage and penetrance requirments, \
some situations with only 1 minor allele read can be used to call het \
genotype but that has low confidence

## Post-Calling Filtering 
1. Generate and filter based on probability-based genotype score
  * `get_adjusted_geno_scores` function
  * Assigns score based on binomial distribution-related relative probability \
of the genotypes based on read ratios, estimated sequencing error, and \
estimated minimum allele penetrance
  * Adjusts the genotype score based on genotypes in other samples at the SV
    * If sampA has low-score genotype but another sample has the same \
genotype and with a high score, then the genotype score in sampA is increased
2. Remove SVs in which no sample has a desired coverage
  * `find_SVs_below_ss_cov` function
  * Example: may set minimum coverage at 10, but want at least one sample \
to have coverage of at least 20
3. Remove SV's with NA genotypes
  * `filt_geno_mat` function
  * can adjust the number of NA genotypes allowed in a SV

